src/dict.c

```diff
@@ -123,133 +130,156 @@ int _dictInit(dict *d, dictType *type)
+    d->ht_size_exp = new_ht_size_exp;
+    d->ht_table = new_ht_table;
     d->rehashidx = 0;
     return DICT_OK;
 }
 
 /* return DICT_ERR if expand was not performed */
 int dictExpand(dict *d, unsigned long size) {
-    return _dictExpand(d, size, NULL);
+    return _dictResize(d, size, NULL);
 }
 
 /* return DICT_ERR if expand failed due to memory allocation failure */
 int dictTryExpand(dict *d, unsigned long size) {
     int malloc_failed;
-    _dictExpand(d, size, &malloc_failed);
-    return malloc_failed? DICT_ERR : DICT_OK;
+    _dictResize(d, size, &malloc_failed);
+    return malloc_failed ? DICT_ERR : DICT_OK;
 }
 
-/* Performs N steps of incremental rehashing. Returns 1 if there are still
- * keys to move from the old to the new hash table, otherwise 0 is returned.
+/* Performs limited steps of incremental rehashing. Returns 1 if there are still
+ * keys to rehash (based on the new size of the HT). Otherwise, 0 is returned.
  *
- * Note that a rehashing step consists in moving a bucket (that may have more
- * than one key as we use chaining) from the old to the new hash table, however
- * since part of the hash table may be composed of empty spaces, it is not
- * guaranteed that this function will rehash even a single bucket, since it
- * will visit at max N*10 empty buckets in total, otherwise the amount of
- * work it does would be unbound and the function may block for a long time. */
-int dictRehash(dict *d, int n) {
-    int empty_visits = n*10; /* Max number of empty buckets to visit. */
+ * Note that a rehashing step consists in updating a bucket (that may have more
+ * than one key as we use chaining). However, since part of the hash table may be
+ * composed of empty buckets, it is not guaranteed that this function will rehash
+ * even a single bucket, since it will visit at max bucket_visits*10 empty
+ * buckets in total, otherwise the amount of work it does would be unbound and
+ * the function may block for a long time. */
+int dictRehash(dict *d, int bucket_visits) {
+    /* Max number of empty buckets to visit. */
+    int empty_visits = bucket_visits * 10;
+
     if (!dictIsRehashing(d)) return 0;
 
-    while(n-- && d->ht_used[0] != 0) {
-        dictEntry *de, *nextde;
+    /* Iterate and rehash all buckets up to the previous size of the hash table,
+     * and rehash all entries in buckets to their updated buckets */
+    long ht_size = DICTHT_SIZE(d->ht_size_exp);
+    long ht_size_prev = DICTHT_SIZE(d->ht_size_exp_prev);
+
+    while (bucket_visits > 0) {
+        dictEntry *entries_to_rehash = d->ht_table[d->rehashidx];
 
-        /* Note that rehashidx can't overflow as we are sure there are more
-         * elements because ht[0].used != 0 */
-        assert(DICTHT_SIZE(d->ht_size_exp[0]) > (unsigned long)d->rehashidx);
-        while(d->ht_table[0][d->rehashidx] == NULL) {
-            d->rehashidx++;
+        /* If HT grows, need to set to NULL current bucket and corresponding new ones
```

Comment:
```suggestion
        /* If HT grows, need to set NULL to current bucket and corresponding new ones
```

Reply:
-> "set current bucket to NULL" even better?

Reply:
Rephrased. Thanks.